### PR TITLE
Separate target specific cmake options into target specific CMakeLists

### PR DESCRIPTION
**Pull request by mathiaswagner**
This PR separates the target(-architecture) specific part mostly to the target specific `CMakeLists.txt`. 
Idea is that apart from setting the target we have no (or in a first iteration almost no) target specific stuff in the non-target specific CMakeLists.txt. 
Intention is to avoid cluttering the CMakeLists and achieve s similar level of separation as we have with the code.

### CODE REVIEW

**Comment by mathiaswagner** [comment]
ping

**Review by maddyscientist — APPROVED** [review]
This looks like a really nice cleanup, and will be critical as more targets are added.

@bjoo this will make your next merge develop more involved, but will be well worth the effort 😉 

@mathiaswagner one question: did you verify Jitify is still ok?

**Comment by bjoo** [comment]
Hi All! This is totally well worth it!!! Thanks for merging!
Best,
B

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: maddyscientist ***@***.***>
Reply-To: lattice/quda ***@***.***>
Date: Wednesday, December 8, 2021 at 12:49 PM
To: lattice/quda ***@***.***>
Cc: Balint Joo ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [lattice/quda] Separate target specific cmake options into target specific CMakeLists (PR #1206)
Resent-From: ***@***.***>
Resent-Date: Wednesday, December 8, 2021 at 12:49 PM


@maddyscientist approved this pull request.

This looks like a really nice cleanup, and will be critical as more targets are added.

@bjoo<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_bjoo&d=DwMFaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=a2u70Wi852kZMX2Ww7pJY-puA_qRlNvLYh60adRS1v09pO__ymtkMcbkIhaPf3O6&s=HJ9d1hzGkNTCZTYP6rBpQvjpJgOuNIt6avVdOBuIFHg&e=> this will make your next merge develop more involved, but will be well worth the effort 😉

@mathiaswagner<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_mathiaswagner&d=DwMFaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=a2u70Wi852kZMX2Ww7pJY-puA_qRlNvLYh60adRS1v09pO__ymtkMcbkIhaPf3O6&s=doy_ghCo1IVdWTdPhy-KkYneqOI_Bf-kBhVca9cXKRQ&e=> one question: did you verify Jitify is still ok?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1206-23pullrequestreview-2D826746991&d=DwMFaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=a2u70Wi852kZMX2Ww7pJY-puA_qRlNvLYh60adRS1v09pO__ymtkMcbkIhaPf3O6&s=SB4zhgJtDqg9DJrFJO41gENwB2e6L5irM6vX2dRywGc&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2BCL5HP5CYJG7PFHN3UP6LCHANCNFSM5HLTIKIA&d=DwMFaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=a2u70Wi852kZMX2Ww7pJY-puA_qRlNvLYh60adRS1v09pO__ymtkMcbkIhaPf3O6&s=XTa2dXgNK9esjIiCaZJtxCfUq6VJ6clJCnUUkkdYPp0&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMFaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=a2u70Wi852kZMX2Ww7pJY-puA_qRlNvLYh60adRS1v09pO__ymtkMcbkIhaPf3O6&s=_rhYU7WA8qp8YIFp_Tn1IR5fsxguSU7cb6kRnNf1iPU&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMFaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=a2u70Wi852kZMX2Ww7pJY-puA_qRlNvLYh60adRS1v09pO__ymtkMcbkIhaPf3O6&s=eLU4h6QK-dLizWYaxu6IDaKNkz3BHMrX6TaPFlDZaGc&e=>.

**Comment by maddyscientist** [comment]
@bjoo this isn't merged yet, it's awaiting your approval 😉

**Review by bjoo — COMMENTED** [review]
Would it be a better factoring if these features would be moved into a directory called  quda/cmake  (as .cmake files) which would then be include in e.g  targets/cuda/CMakeLists.txt? so that they would be factored from the build of the target specific code?  I ask because there are also .cu files in just lib/ that could include these too...?

**Comment by mathiaswagner** [comment]
@maddyscientist I still need to verify jitify.

@bjoo let me think about that. Including the target specific part in the `lib/CMakeLists.txt` clearly has its benefits. Location-wise I like having it in the target directory but I like the `include`.

**Comment by mathiaswagner** [comment]
@bjoo I switched to using `include()` instead of `add_subdirectoty()`. Difference is not huge but makes some things a little cleaner. I like that.

**Comment by mathiaswagner** [comment]
@maddyscientist jitify seems to work (compiled and ran dslash_test).

**Comment by maddyscientist** [comment]
@bjoo are you happy now with this?

**Comment by mathiaswagner** [comment]
ping

File: lib/CMakeLists.txt [hunk 1]
```diff
@@ -153,183 +154,74 @@ endif()
...
+
+# add target specific files
+if(${QUDA_TARGET_TYPE} STREQUAL "CUDA")
+  include(targets/cuda/target_cuda.cmake)
+endif()
+if(${QUDA_TARGET_TYPE} STREQUAL "HIP")
+  add_subdirectory(targets/hip)
+  target_include_directories(quda PRIVATE ../include/targets/hip)
+  set(QUDA_TARGET_HIP ON)
 endif()
```
Comment (bjoo) [inline lib/CMakeLists.txt:196]: I will need to make the HIP target similar to the CUDA target here, but that will be my problem...

**Review by bjoo — APPROVED** [review]
THis all looks great. Magma is gone, and all the target stuff in in lib/targets/ARCH/target_ARCH.cmake which is nice. I have to make my HIP backend fit, but I should have a good model. This is a nice refactoring and makes the other makefiles a lot tidyer. Thanks @mathiaswagner for your hard work on this.

**Pull request merged**